docs/aztec/protocol/trees/indexed-merkle-tree.mdx

```diff
@@ -85,85 +118,218 @@ Suppose we want to show that the value `20` doesn't exist in the tree. We just r
 
-#### Number of non-membership proof constraints, in total:
+This is already a massive performance improvement, however we can go further, as this tree is not sparse. We can perform batch insertions.
 
-- `n` hashes of 2 field elements (where `n` is the height of the tree).
-- `1` hash of 3 field elements.
-- `2` range checks.
-- A handful of equality constraints.
+## Batch insertions
 
-### Doing a non-membership check, then an insertion of that value
+As our nullifiers will all be inserted deterministically (append only), we can insert entire subtrees into our tree rather than appending nodes one by one, this optimization is globally applied to append only merkle trees. I wish this was it, but for every node we insert, we must also update low nullifier pointers, this introduces a bit of complexity while performing subtree insertions, as the low nullifier itself may exist within the subtree we are inserting - we must be careful how we prove these sort of insertions are correct (addressed later).
+We must update all of the impacted low nullifiers before.
 
-We combine the logic of the above two sections, where `oldLeaf` is renamed to `loLeaf` throughout, and `nonLeaf` is renamed to `newLeaf`. We can actually make some savings.
+First we will go over batch insertions in an append only merkle tree.
 
-- Hash the low leaf.
-- Prove the low leaf exists in the tree: `n` hashes.
-- Check the new leaf 'would have' belonged in the range given by the low leaf: `2` range checks.
-  - If (`loLeaf.nextInd == 0`):
-    - Special case, the low leaf is at the very end, so the non leaf must be higher than all values in the tree:
-    - `assert(loLeaf.val < newLeaf.val)`
-  - Else:
-    - `assert(loLeaf.val < newLeaf.val)`
-    - `assert(loLeaf.nextVal > newLeaf.val)`
-- Copy over the old leaf's `next` values to the new leaf:
-  - `assert(oldLeaf.nextInd == newLeaf.nextInd)`
-  - `assert(oldLeaf.nextVal == newLeaf.nextVal)`
-- Update the old leaf to point to the new leaf, to create `updatedLeaf`.
-  - `assert(updatedLeaf.nextInd == newLeaf.ind)`
-  - `assert(updatedLeaf.nextVal == newLeaf.val)`
-- Replace the `updatedLeaf` in the tree: `n` hashes.
-- Hash the new leaf.
-- Add the `newLeaf` to the tree: `n` hashes.
-
-#### Number of constraints, in total:
+1. First we prove that the subtree we are inserting into consists of all empty values.
+1. We work out the root of an empty subtree, and perform an inclusion proof for an empty root, which proves that there is nothing within our subtree.
+1. We re-create our subtree within our circuit.
+1. We then use the same sibling path the get the new root of the tree after we insert the subtree.
 
-- `3n` hashes of 2 field elements (where `n` is the height of the tree).
-- `3` hashes of 3 field elements.
-- `2` range checks.
-- A handful of equality constraints.
+In the following example we insert a subtree of size 4 into our tree at step 4. above. Our subtree is greyed out as it is "pending".
```

Comment:
Would be good with a note on the colors outside the grey, e.g., the orange and gree. Also there seem to be used 2 different greens throughout the steps.

Reply:
two greens was an error